src/game/client/draw/w3dtankdraw.cpp

```diff
@@ -0,0 +1,376 @@
+        { nullptr, nullptr, nullptr, 0 },
+    };
+    // clang-format on
+
+    W3DModelDrawModuleData::Build_Field_Parse(p);
+    p.Add(dataFieldParse, 0);
+}
+
+W3DTankDraw::W3DTankDraw(Thing *thing, ModuleData const *module_data) :
+    W3DModelDraw(thing, module_data),
+    m_treadDebrisLeft(nullptr),
+    m_treadDebrisRight(nullptr),
+    m_prevRenderObj(nullptr),
+    m_treadCount(0)
+{
+    for (int i = 0; i < MAX_TREADS_PER_TANK; i++) {
+        m_treads[i].m_robj = nullptr;
+    }
+
+    m_treadDirection.x = 1.0f;
+    m_treadDirection.y = 0.0f;
+    m_treadDirection.z = 0.0f;
+    Create_Emitters();
+}
+
+W3DTankDraw::~W3DTankDraw()
+{
+    // BUGFIX: Free emitters to avoid a leak
+    Toss_Emitters();
+
+    for (int i = 0; i < MAX_TREADS_PER_TANK; i++) {
+        Ref_Ptr_Release(m_treads[i].m_robj);
+    }
+}
+
+NameKeyType W3DTankDraw::Get_Module_Name_Key() const
+{
+    static const NameKeyType nk = g_theNameKeyGenerator->Name_To_Key("W3DTankDraw");
+    return nk;
+}
+
+void W3DTankDraw::Do_Draw_Module(const Matrix3D *transform)
+{
+    bool frozen = g_theTacticalView->Is_Time_Frozen() && !g_theTacticalView->Is_Camera_Movement_Finished();
+    frozen = frozen || g_theScriptEngine->Is_Time_Frozen_Debug() || g_theScriptEngine->Is_Time_Frozen_Script();
+
+    if (!frozen && Get_Render_Object() != nullptr) {
+        if (Get_Render_Object() != m_prevRenderObj) {
+            Update_Tread_Objects();
+        }
+
+        Object *object = Get_Drawable()->Get_Object();
+
+        if (object != nullptr) {
+            PhysicsBehavior *physics = object->Get_Physics();
+
+            if (physics) {
+                const Coord3D &velocity = physics->Get_Velocity();
+                int len = velocity.x * velocity.x + velocity.y * velocity.y;
+
+                if (len > 0.00001f && !Get_Drawable()->Is_Hidden() && !Get_Fully_Obscured_By_Shroud()) {
+                    Start_Move_Debris();
+                } else {
+                    Stop_Move_Debris();
+                }
+
+                len = GameMath::Sqrt(len);
+                Coord3D mul;
+                mul.x = 0.5f * len + 0.1f;
+
+                if (mul.x > 1.0f) {
+                    mul.x = 1.0f;
+                }
+
+                mul.y = mul.x;
+                mul.z = len + 0.1f;
+
+                if (mul.z > 1.0f) {
+                    mul.z = 1.0f;
+                }
+
+                m_treadDebrisLeft->Set_Velocity_Multiplier(&mul);
+                m_treadDebrisRight->Set_Velocity_Multiplier(&mul);
+                m_treadDebrisLeft->Set_Burst_Count_Multiplier(mul.z);
+                m_treadDebrisRight->Set_Burst_Count_Multiplier(mul.z);
+
+                if (m_treadCount != 0) {
+                    PhysicsTurningType turning = physics->Get_Turning();
+                    float rate = Get_W3D_Tank_Draw_Module_Data()->m_treadAnimationRate;
+                    float speed = object->Get_AI_Update_Interface()->Get_Cur_Locomotor_Speed();

```

Comment:
Can speed be 0 ? If so we would divide by 0 further down.

Reply:
No it isn't a problem.